# Analyzing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

The data dictionary provided with data is as follows:

- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which which year the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which which year the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import pandas as pd

#Try without specifying any encoding (which will default to UTF-8)
#If you get an encoding error, try the next two most popular encodings 
#(Latin-1 and Windows-1252) until you are able to read the file without error.
autos = pd.read_csv('autos.csv', encoding='Latin-1') 
autos.head()


dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

From the above information, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.


In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

# Cleaning column names

In [2]:
print('Original CamelCase style: ')
print(autos.columns)
print('\n')

column_names = {
    'yearOfRegistration': 'registration_year',
'monthOfRegistration': 'registration_month',
'notRepairedDamage': 'unrepaired_damage',
    'vehicleType': 'vehicle_type',
'dateCreated': 'ad_created',
'dateCrawled': 'date_crawled',
'offerType': 'offer_type',
'powerPS': 'power_ps',
'fuelType': 'fuel_type',
'nrOfPictures': 'pictures_nr',
'postalCode': 'postal_code',
'lastSeen': 'last_seen'}

autos.rename(columns=column_names, inplace=True)
print('Updated to snakecase style: ')
print(autos.columns)

print('\n')
print(autos.head())

Original CamelCase style: 
Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Updated to snakecase style: 
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_nr', 'postal_code',
       'last_seen'],
      dtype='object')


          date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"

# Initial exploration and cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [3]:
print(autos.describe(include='all'))

               date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
c


Our initial observations:

- There are a number of text columns where all (or nearly all) of the values are the same:
  - seller
  - offer_type

- The *pictures_nr* column looks odd, we'll need to investigate this further.

In [4]:
autos['pictures_nr'].value_counts()

0    371528
Name: pictures_nr, dtype: int64

It looks like the *pictures_nr* column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value

In [5]:
autos.drop(['seller', 'offer_type', 'pictures_nr'], axis=1, inplace=True)

print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 17 columns):
date_crawled          371528 non-null object
name                  371528 non-null object
price                 371528 non-null int64
abtest                371528 non-null object
vehicle_type          333659 non-null object
registration_year     371528 non-null int64
gearbox               351319 non-null object
power_ps              371528 non-null int64
model                 351044 non-null object
kilometer             371528 non-null int64
registration_month    371528 non-null int64
fuel_type             338142 non-null object
brand                 371528 non-null object
unrepaired_damage     299468 non-null object
ad_created            371528 non-null object
postal_code           371528 non-null int64
last_seen             371528 non-null object
dtypes: int64(6), object(11)
memory usage: 48.2+ MB
None


# Exploring the kilometer and price columns

In [6]:
def explor_data(series):
    print('Number of unique values', series.unique().shape)
    print('\n')
    print('View min/max/median/mean etc', series.describe())
    print('\n')
    print('View the highest and lowest values with their counts: ')
    print(series.value_counts().sort_index().head(20))
    
explor_data(autos['kilometer'])
print('\n')
explor_data(autos['price'])

Number of unique values (13,)


View min/max/median/mean etc count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64


View the highest and lowest values with their counts: 
5000        7069
10000       1949
20000       5676
30000       6041
40000       6376
50000       7615
60000       8669
70000       9773
80000      11053
90000      12523
100000     15920
125000     38067
150000    240797
Name: kilometer, dtype: int64


Number of unique values (5597,)


View min/max/median/mean etc count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


View the highest and lowest values with their counts: 
0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8    


We can see that the values in the *kilometer* field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

Again, the prices in this column seem rounded, however given there are 5597 unique values in the column, that may just be people's tendency to round prices on the site.

There are 10778 cars listed with $0 price - given that this is only 3% of the of the cars, we might consider removing these rows. The maximum price is more than 20 hundred million dollars, which seems a lot, let's look at the highest prices further.

In [7]:
print(autos['price'].value_counts().sort_index(ascending=False).head(20))

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64



There are a number of listings with prices below \$20, including about 11000 at \$0. There are also a small number of listings with very high values, including 51 at around or over $10 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$3895000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [8]:
autos = autos[autos['price'].between(1,3895000)]
autos['price'].describe()

count    3.606980e+05
mean     6.068401e+03
std      1.818741e+04
min      1.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.895000e+06
Name: price, dtype: float64

# Exploring the date columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:


In [9]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [10]:
import datetime as dt

column_time = ['date_crawled','ad_created','last_seen']

for col in column_time:
    autos[col] = pd.to_datetime(autos[col])
    autos[col + '_date'] = autos[col].dt.date
    
autos[['date_crawled_date','ad_created_date','last_seen_date']][0:5]
# autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled_date ad_created_date last_seen_date
0        2016-03-24      2016-03-24     2016-04-07
1        2016-03-24      2016-03-24     2016-04-07
2        2016-03-14      2016-03-14     2016-04-05
3        2016-03-17      2016-03-17     2016-03-17
4        2016-03-31      2016-03-31     2016-04-06

In [15]:
column_date = ['date_crawled_date','ad_created_date','last_seen_date']

for col in column_date:
    print(col)
    print(autos[col].value_counts(normalize=True, dropna=False).sort_index())
    print('\n')

date_crawled_date
2016-03-05    0.025545
2016-03-06    0.014480
2016-03-07    0.035653
2016-03-08    0.033466
2016-03-09    0.034112
2016-03-10    0.032648
2016-03-11    0.032775
2016-03-12    0.036244
2016-03-13    0.015781
2016-03-14    0.036327
2016-03-15    0.033424
2016-03-16    0.030200
2016-03-17    0.031655
2016-03-18    0.013119
2016-03-19    0.035268
2016-03-20    0.036399
2016-03-21    0.035695
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032798
2016-03-26    0.031969
2016-03-27    0.030230
2016-03-28    0.035060
2016-03-29    0.034123
2016-03-30    0.033529
2016-03-31    0.031874
2016-04-01    0.034150
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037630
2016-04-05    0.012786
2016-04-06    0.003127
2016-04-07    0.001619
Name: date_crawled_date, dtype: float64


ad_created_date
2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
2015-08-10    0.000003
2015

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest more than 1 year.

In [17]:
autos['registration_year'].describe()

count    360698.000000
mean       2004.451877
std          82.110038
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

# Dealing with incorrect registration year data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [18]:
(~autos['registration_year'].between(1900, 2016)).sum()/autos.shape[0]

0.03875264071328369

Given that this is less than 4% of our data, we will remove these rows.

In [21]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=False).head(10)

2016    0.026612
2015    0.008430
2014    0.013755
2013    0.017657
2012    0.026993
2011    0.034599
2010    0.035392
2009    0.044702
2008    0.046248
2007    0.050490
Name: registration_year, dtype: float64

# Exploring price by brand

In [35]:
brand_counts = autos['brand'].value_counts(normalize=True)
print(brand_counts)

volkswagen        0.211675
bmw               0.109870
opel              0.106397
mercedes_benz     0.096833
audi              0.089536
ford              0.068912
renault           0.047508
peugeot           0.030148
fiat              0.025686
seat              0.018658
skoda             0.015684
mazda             0.015381
smart             0.014329
citroen           0.013948
nissan            0.013596
toyota            0.012930
hyundai           0.009971
sonstige_autos    0.009547
mini              0.009382
volvo             0.009146
mitsubishi        0.008234
honda             0.007531
kia               0.006913
suzuki            0.006365
alfa_romeo        0.006308
porsche           0.006259
chevrolet         0.005024
chrysler          0.003862
dacia             0.002495
jeep              0.002195
land_rover        0.002166
daihatsu          0.002160
subaru            0.002117
jaguar            0.001736
saab              0.001465
daewoo            0.001457
trabant           0.001407
l

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

In [44]:
brand_price = {}
common_brands = brand_counts[brand_counts > 0.05].index

for brand in common_brands:
    brand_price[brand] = autos.loc[autos['brand'] == brand, 'price'].mean()
    print(brand, brand_price[brand])



volkswagen 5452.33870176586
bmw 8573.86244552948
opel 3028.965952832746
mercedes_benz 8609.055965926014
audi 9166.50879396985
ford 3779.8030385468546



Of the top 6 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.


In [47]:
brand_mean_prices = pd.Series(brand_price).sort_values(ascending=False)
# print(brand_mean_prices.sort_values(ascending=False))
df_brand_mean_prices = pd.DataFrame(brand_mean_prices, columns=['mean_price'])
df_brand_mean_prices

mean_price
audi           9166.508794
mercedes_benz  8609.055966
bmw            8573.862446
volkswagen     5452.338702
ford           3779.803039
opel           3028.965953

# Exploring Mileage

In [50]:
mean_mileage = {}
for brand in common_brands:
    mean_mileage[brand] = autos.loc[autos['brand'] == brand, 'kilometer'].mean()
    print(brand, mean_mileage[brand])
    
brand_mean_mileage = pd.Series(mean_mileage).sort_values(ascending=False)
df_brand_mean_mileage = pd.DataFrame(brand_mean_mileage, columns=['mean_mileage'])
df_brand_mean_mileage

volkswagen 128387.22204054937
bmw 132791.77823279257
opel 128721.06261859582
mercedes_benz 130566.80764877584
audi 129440.7937121505
ford 123663.41606328213


mean_mileage
bmw            132791.778233
mercedes_benz  130566.807649
audi           129440.793712
opel           128721.062619
volkswagen     128387.222041
ford           123663.416063

In [51]:
brand_info = pd.concat([df_brand_mean_prices, df_brand_mean_mileage], axis=1)
brand_info

mean_price   mean_mileage
audi           9166.508794  129440.793712
bmw            8573.862446  132791.778233
ford           3779.803039  123663.416063
mercedes_benz  8609.055966  130566.807649
opel           3028.965953  128721.062619
volkswagen     5452.338702  128387.222041

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.